include/openssl/pkcs12.h.in

```diff
@@ -221,15 +237,25 @@ int PKCS12_parse(PKCS12 *p12, const char *pass, EVP_PKEY **pkey, X509 **cert,
 PKCS12 *PKCS12_create(const char *pass, const char *name, EVP_PKEY *pkey,
                       X509 *cert, STACK_OF(X509) *ca, int nid_key, int nid_cert,
                       int iter, int mac_iter, int keytype);
+PKCS12 *PKCS12_create_ex(EVP_PKEY *pkey, X509 *cert, STACK_OF(X509) *ca,
+                         OSSL_PARAM[] params, OSSL_LIB_CTX *ctx, const char *propq);
 
 PKCS12_SAFEBAG *PKCS12_add_cert(STACK_OF(PKCS12_SAFEBAG) **pbags, X509 *cert);
 PKCS12_SAFEBAG *PKCS12_add_key(STACK_OF(PKCS12_SAFEBAG) **pbags,
                                EVP_PKEY *key, int key_usage, int iter,
                                int key_nid, const char *pass);
+PKCS12_SAFEBAG *PKCS12_add_key_ex(STACK_OF(PKCS12_SAFEBAG) **pbags,
```

Comment:
With this one and the PKCS7 stack one below, what would the libctx be used for.. Potentially each object in the stack already has an associated libctx, propq (i.e PKCS7 stores this internally, the same may apply to the safebags?)

Reply:
PKCS12_add_key() can be called with an empty (NULL) stack, in which case there would be no way to get the libctx. Safer to provide one explicitly.

For PKCS12_add_safe() it may be possible to retrieve the libctx from the first bag in the list as (I would assume) there would always be at least one bag that we are encapsulating in the P7 data. It would first need to be stored in the PKCS12_SAFEBAG_st structure, so that would mean also passing it into PKCS12_add_cert() and PKCS12_add_secret()? I'm not sure that would be any better.